src/libraries/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/DontBeACA.cs

```diff
@@ -0,0 +1,484 @@
+
+                byte[] serial = { 1, 1, 2, 3, 5, 8, 13, 21 };
+                DateTimeOffset notBefore = DateTimeOffset.UtcNow.AddMinutes(-2);
+                DateTimeOffset notAfter = notBefore.AddMinutes(20);
+
+                using (X509Certificate2 issued = req.Create(issuerCert, notBefore, notAfter, serial))
+                {
+                    Assert.NotNull(issued);
+                    AssertExtensions.SequenceEqual(serial, issued.SerialNumberBytes.Span);
+                    AssertExtensions.SequenceEqual(reqName.RawData, issued.SubjectName.RawData);
+                    AssertExtensions.SequenceEqual(issuerCert.SubjectName.RawData, issued.IssuerName.RawData);
+                    Assert.Equal(notBefore.DateTime, issued.NotBefore.ToUniversalTime(), TimeSpan.FromSeconds(1));
+                    Assert.Equal(notAfter.DateTime, issued.NotAfter.ToUniversalTime(), TimeSpan.FromSeconds(1));
+
+                    X509ExtensionCollection extensions = issued.Extensions;
+                    Assert.Equal(7, extensions.Count);
+
+                    // Extensions are a SEQUENCE OF, not a SET OF, so the order the CA wrote them is the order they appear.
+                    Assert.IsType<X509SubjectKeyIdentifierExtension>(extensions[0]);
+                    Assert.IsType<X509BasicConstraintsExtension>(extensions[1]);
+                    Assert.IsType<X509AuthorityKeyIdentifierExtension>(extensions[2]);
+                    Assert.Equal("2.5.29.31", extensions[3].Oid.Value);
+                    Assert.IsType<X509AuthorityInformationAccessExtension>(extensions[4]);
+                    Assert.IsType<X509KeyUsageExtension>(extensions[5]);
+                    Assert.IsType<X509SubjectAlternativeNameExtension>(extensions[6]);
+                }
+            }
+
+            static CertificateRequest IngestRequest(byte[] pkcs10, X509Certificate2 issuerCert)
+            {
+                CertificateRequest req = CertificateRequest.LoadCertificateRequest(
+                    pkcs10,
+                    HashAlgorithmName.SHA256,
+                    out int bytesConsumed,
+                    skipSignatureValidation: true,
+                    signerSignaturePadding: RSASignaturePadding.Pss);
+
+                // TODO: How to validate the Subject?
+
+                AsnEncodedData? challengePassword = null;
+
+                foreach (AsnEncodedData otherAttribute in req.OtherRequestAttributes)
+                {
+                    if (otherAttribute?.Oid?.Value == "1.2.840.113549.1.9.7")
+                    {
+                        if (challengePassword is not null)
+                        {
+                            throw new InvalidOperationException("Two challenge passwords provided");
+                        }
+
+                        challengePassword = otherAttribute;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException(
```

Comment:
This is _so_ important that I wonder if the API should make it not optional to skip this.  The same for extensions.  If you fail to handle one (or even understand one), you _really_ shouldn't be signing the thing.  I'm not sure if it's cool to say "pit of success" anymore, but this seems like it's too easy to skip something.  It almost feels like the thing that you are signing and the thing you are loading maybe shouldn't be the same object... forcing you to at least _copy_ each thing (which you should be validating as you copy, since there are, to my knowledge, no things that a CA should just blindly copy over to a signed cert).

I could be wrong, and maybe the normal case for an attribute/extension is "blindly copy/sign"... but in my head that sounds scary.

Reply:
Maybe you load a request and create a "ToBeSignedCertificate" or somethiing... though that's a terrible name.  Uh... "UnsignedCertificate"?

Reply:
Well, attributes are... weird.  They're just sort of extra messages you're passing to the CA, like the challenge password.

In fact, RFC https://datatracker.ietf.org/doc/html/rfc2985#page-16 only defines 3 attributes:

* Challenge Password
* Extension Request (gets mapped in special, and why the attributes bucket is called "Other" Request Attributes
* Extended Certificate Attributes (deprecated) (like Extension Request, but to a PKCS#6 certificate, which are deprecated)

It's an open ended bucket because that's what the spec says, and custom protocols might define one.

So... for attributes (this bucket) it's less critical.

Reply:
Ah, so if you sort of ignore this one, it would be fine, since you are signing the attributes? They are, essentially attributes of the "request" not of the "certificate"?

Reply:
> They are, essentially attributes of the "request" not of the "certificate"?

Aye.  Except the Extension Request Attribute, which gets mapped into the CertificateExtensions collection :smile:.

Reply:
Some thoughts for the extensions-are-scary problem:

* @vcsjones suggested adding an OidCollection into Load, and that restricts what extensions can get loaded
* The spiritual equivalent of an "I have read and agree" checkbox: Add `public bool ExtensionsValidated { get; set; }`, it defaults to `true`, except from Load, when it defaults to `false`, and you have to set it to true before calling sign.
* Don't populate CertificateExtensions, just manifest that attribute into a CertificateExtensionsAttribute (name needs work) that has its own X509ExtensionCollection and you can copy things over.
* Open to other ideas.

Reply:
I'd worry that the "OidCollection" would just be "I am capable of validating these", and it would still be easy to _fail_ to validate them.  You added an Oid to the Load, and just forgot to at it to your for-switch loop of validation.  Or the reverse, you added it to the loop, but missed it in the Load call, so it just never got loaded.  It seems like forcing the "copy" means the "I expect this" and "I validated this" are one-in-the-same, so it's harder to miss a piece of it.

If the collection was something else, like a Dictionary<Oid,Predicate<Extension>> mapping, that could work too if that's a cheaper alternative.  Or some sort of "IExtensionValidator", since dictionary mapping things to functions is a sort of hacky version of an interface.

Reply:
To be honest I'm not sure `CertificateRequest` is the right place for that kind of validation if we want something like that then. I also think these APIs are fundamentally a laser cannon pointed at ones foot. _Some_ safety mechanisms are probably a good idea - but building out a validation infrastructure seems like a lot of effort for something that, no matter what, people need to be **very** certain what they are doing.

Reply:
If we're that distrustworthy of people loading extensions, then I would consider:

```C#
CertificateRequest.Load(..., bool unsafeLoadCertificateExtensionRequests = false);
```